server/src/test/java/org/elasticsearch/cluster/routing/RoutingTableTests.java

```diff
@@ -392,6 +398,39 @@ public void testAddAsRecovery() {
         }
     }
 
+    public void testRemoveReplicasInRightOrder() {
+        final List<ShardRoutingState> rightRemoveOrder = List.of(UNASSIGNED, INITIALIZING, RELOCATING, STARTED);
+        Index index = new Index("index", "uuid");
+        ShardId shardId = new ShardId(index, 0);
+        List<ShardRouting> shards = new ArrayList<>();
+        shards.add(TestShardRouting.newShardRouting(shardId, "node1", true, STARTED));
+        shards.add(TestShardRouting.newShardRouting(shardId, null, false, UNASSIGNED));
+        shards.add(TestShardRouting.newShardRouting(shardId, "node2", false, INITIALIZING));
+        shards.add(TestShardRouting.newShardRouting(shardId, "node3", false, STARTED));
+        shards.add(TestShardRouting.newShardRouting(shardId, "node4", "node5", false, ShardRoutingState.RELOCATING));
+
+        for (int removeReplicaNumber = 0; removeReplicaNumber <= rightRemoveOrder.size(); removeReplicaNumber++) {
+            IndexRoutingTable.Builder builder = new IndexRoutingTable.Builder(index);
+            Randomness.shuffle(shards);
+            for (ShardRouting shard: shards) {
+                builder.addShard(shard);
+            }
+            for (int round = 0; round < removeReplicaNumber; round++) {
+                builder.removeReplica();
+            }
+            IndexRoutingTable indexRoutingTable = builder.build();
+
+            for (ShardRoutingState state: rightRemoveOrder) {
+                int theRoundThisStateShouldBeRemoved = rightRemoveOrder.indexOf(state) + 1;
+                if (removeReplicaNumber < theRoundThisStateShouldBeRemoved) {
+                    assertThat(indexRoutingTable.shard(0).replicaShardsWithState(state).size(), equalTo(1));
```

Comment:
A message on failure would help here, and in the other assertion too. How about this?

```suggestion
                    assertThat(indexRoutingTable.prettyPrint(), indexRoutingTable.shard(0).replicaShardsWithState(state).size(), equalTo(1));
```